core/shared/src/main/scala/zio/internal/macros/LayerBuilder.scala

```diff
@@ -132,6 +129,79 @@ final case class LayerBuilder[Type, Expr](
     }
   }
 
+  /**
+   * Returns tree representing final `ZLayer[In, Err, Out]`
+   *
+   * ==If [[method]] is not [[ProvideMethod.ProvideSomeShared]]==
+   *   - `In` will be some autogenerated type
+   *   - `Err` will be the union of all errors in the provided layers
+   *   - `Out` will be AND'ed together types from [[target0]]
+   *
+   * ==If [[method]] is [[ProvideMethod.ProvideSomeShared]]==
+   *   - `In` will be some autogenerated type
+   *   - `Err` will be the union of all errors in the provided layers
+   *   - `Out` will be some autogenerated type
+   *
+   * And following relation exists: `In & Out` <:< AND'ed together types from
+   * [[target0]]
+   */
+  def buildAuto: Expr = {
+    assertNoAmbiguity()
+
+    val remainderTypeFactory = (t: Type) => {
+      Some(typeToNode(t))
+    }
+
+    val nodes: List[Node[Type, Expr]] = providedLayerNodes ++ sideEffectNodes
+    val nodesOutputs                  = nodes.flatMap(n => n.outputs)
+    val graph                         = Graph(nodes, typeEquals, remainderTypeFactory)
+
+    // ProvideSomeShared ignores optionalInputs - they are the "shared" part
+    val (requiredInputs, optionalInputs) = target0
+      .partition(t => nodesOutputs.exists(tpe => typeEquals(tpe, t)))
+
+    /**
+     * Build the layer tree. This represents the structure of a successfully
+     * constructed ZLayer that will build the target types. This, of course, may
+     * fail with one or more GraphErrors.
+     */
+    val layerTreeEither: Either[::[GraphError[Type, Expr]], LayerTree[Expr]] =
+      for {
+        inputs <- if (method.isProvideSomeShared) {
+                    graph.buildComplete(requiredInputs)
+                  } else {
+                    graph.buildComplete(requiredInputs ++ optionalInputs)
+                  }
+        sideEffects <- graph.buildNodes(sideEffectNodes)
+      } yield sideEffects ++ inputs
+
+    layerTreeEither match {
+      case Left(buildErrors) =>
+        reportBuildErrors(buildErrors)
+
+      case Right(tree) =>
+        // TODO: Write some tests for previous implementation and prepare new implementation based on them
+//        warnUnused(tree)
```

Comment:
Should this be commented? I have a feeling like it's needed

Reply:
I'm not sure whether we have a test to check that a warning is raised in case there are unused dependencies provided. Could you double-check whether there is a test or that the warnings are working as intended?

Reply:
![image](https://github.com/user-attachments/assets/fbe39441-f13c-4153-a948-afcfaa4b9a4f)

I don't think there are tests for it yet.

The issue is that warnUnused is using `remainderNodes` field which is not correct for autos

Reply:
I'm not sure I understand your comment / screenshot. My question was whether we still get a warning if we provide more than it's required by the environment.

Reply:
I'm very suprised it works correctly when uncommenting warnUnused

![image](https://github.com/user-attachments/assets/8f649083-d3d2-44f3-8ffe-4ef050efae7c)

I guess I will uncomment it - it only affects `auto`s - if some bug happen I will try to write tests for them and make this unused stuff checking better.